src/rust/src/sapling.rs

```diff
@@ -0,0 +1,176 @@
+        sighash_value: &[u8; 32],
+    ) -> bool {
+        // Deserialize the value commitment
+        let cv = match de_ct(jubjub::ExtendedPoint::from_bytes(cv)) {
+            Some(p) => p,
+            None => return false,
+        };
+
+        // Deserialize the anchor, which should be an element
+        // of Fr.
+        let anchor = match de_ct(bls12_381::Scalar::from_bytes(anchor)) {
+            Some(a) => a,
+            None => return false,
+        };
+
+        // Deserialize rk
+        let rk = match redjubjub::PublicKey::read(&rk[..]) {
+            Ok(p) => p,
+            Err(_) => return false,
+        };
+
+        // Deserialize the signature
+        let spend_auth_sig = match Signature::read(&spend_auth_sig[..]) {
+            Ok(sig) => sig,
+            Err(_) => return false,
+        };
+
+        // Deserialize the proof
+        let zkproof = match Proof::read(&zkproof[..]) {
+            Ok(p) => p,
+            Err(_) => return false,
+        };
+
+        self.0.check_spend(
+            cv,
+            anchor,
+            nullifier,
+            rk,
+            sighash_value,
+            spend_auth_sig,
+            zkproof,
+            unsafe { SAPLING_SPEND_VK.as_ref() }.unwrap(),
+        )
+    }
+    fn check_output(
+        &mut self,
+        cv: &[u8; 32],
+        cm: &[u8; 32],
+        epk: &[u8; 32],
+        zkproof: &[u8; GROTH_PROOF_SIZE],
+    ) -> bool {
+        // Deserialize the value commitment
+        let cv = match de_ct(jubjub::ExtendedPoint::from_bytes(cv)) {
+            Some(p) => p,
+            None => return false,
+        };
+
+        // Deserialize the commitment, which should be an element
+        // of Fr.
+        let cm = match de_ct(bls12_381::Scalar::from_bytes(cm)) {
+            Some(a) => a,
+            None => return false,
+        };
+
+        // Deserialize the ephemeral key
+        let epk = match de_ct(jubjub::ExtendedPoint::from_bytes(epk)) {
+            Some(p) => p,
+            None => return false,
+        };
+
+        // Deserialize the proof
+        let zkproof = match Proof::read(&zkproof[..]) {
+            Ok(p) => p,
+            Err(_) => return false,
+        };
+
+        self.0.check_output(
+            cv,
+            cm,
+            epk,
+            zkproof,
+            unsafe { SAPLING_OUTPUT_VK.as_ref() }.unwrap(),
+        )
+    }
+    fn final_check(
+        &self,
+        value_balance: i64,
+        binding_sig: &[u8; 64],
+        sighash_value: &[u8; 32],
+    ) -> bool {
+        let value_balance = match Amount::from_i64(value_balance) {
+            Ok(vb) => vb,
+            Err(()) => return false,
+        };
+
+        // Deserialize the signature
+        let binding_sig = match Signature::read(&binding_sig[..]) {
+            Ok(sig) => sig,
+            Err(_) => return false,
+        };
+
+        self.0
+            .final_check(value_balance, sighash_value, binding_sig)
```

Comment:
These APIs (`final_check` and `SaplingVerificationContext::final_check`) also have arguments in an inconsistent order, but here there's no argument that we're preserving the field order in the v4 transaction encoding, because `sighash_value` is not a field.